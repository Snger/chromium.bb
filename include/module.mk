# Copyright (c) 2012 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

include common.mk

install-header: CC_LIBRARY(src/libevdev.so.0)
	install -D -m 0644 include/libevdev/libevdev.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev.h
	install -D -m 0664 include/libevdev/libevdev_event.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_event.h
	install -D -m 0664 include/libevdev/libevdev_mt.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_mt.h
	install -D -m 0664 include/libevdev/libevdev_log.h \
		 $(DESTDIR)/usr/include/libevdev/libevdev_log.h