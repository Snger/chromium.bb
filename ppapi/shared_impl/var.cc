// Copyright (c) 2011 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "ppapi/shared_impl/var.h"

#include <limits>

#include "base/logging.h"
#include "base/string_number_conversions.h"
#include "base/string_util.h"
#include "ppapi/c/pp_var.h"
#include "ppapi/shared_impl/tracker_base.h"

namespace ppapi {

// Var -------------------------------------------------------------------------

Var::Var(PP_Module module) : pp_module_(module), var_id_(0) {
}

Var::~Var() {
}

// static
std::string Var::PPVarToLogString(PP_Var var) {
  switch (var.type) {
    case PP_VARTYPE_UNDEFINED:
      return "[Undefined]";
    case PP_VARTYPE_NULL:
      return "[Null]";
    case PP_VARTYPE_BOOL:
      return var.value.as_bool ? "[True]" : "[False]";
    case PP_VARTYPE_INT32:
      return base::IntToString(var.value.as_int);
    case PP_VARTYPE_DOUBLE:
      return base::DoubleToString(var.value.as_double);
    case PP_VARTYPE_STRING: {
      scoped_refptr<StringVar> string(StringVar::FromPPVar(var));
      if (!string)
        return "[Invalid string]";

      // Since this is for logging, escape NULLs, truncate length.
      std::string result;
      const size_t kTruncateAboveLength = 128;
      if (string->value().size() > kTruncateAboveLength)
        result = string->value().substr(0, kTruncateAboveLength) + "...";
      else
        result = string->value();

      std::string null;
      null.push_back(0);
      ReplaceSubstringsAfterOffset(&result, 0, null, "\\0");
      return result;
    }
    case PP_VARTYPE_OBJECT:
      return "[Object]";
    default:
      return "[Invalid var]";
  }
}

// static
void Var::PluginAddRefPPVar(PP_Var var) {
  if (var.type == PP_VARTYPE_STRING || var.type == PP_VARTYPE_OBJECT) {
    if (!TrackerBase::Get()->AddRefVar(static_cast<int32>(var.value.as_id)))
      DLOG(WARNING) << "AddRefVar()ing a nonexistent string/object var.";
  }
}

// static
void Var::PluginReleasePPVar(PP_Var var) {
  if (var.type == PP_VARTYPE_STRING || var.type == PP_VARTYPE_OBJECT) {
    if (!TrackerBase::Get()->UnrefVar(static_cast<int32>(var.value.as_id)))
      DLOG(WARNING) << "ReleaseVar()ing a nonexistent string/object var.";
  }
}

StringVar* Var::AsStringVar() {
  return NULL;
}

NPObjectVar* Var::AsNPObjectVar() {
  return NULL;
}

int32 Var::GetExistingVarID() const {
  return var_id_;
}

int32 Var::GetOrCreateVarID() {
  TrackerBase* tracker = TrackerBase::Get();
  if (var_id_) {
    if (!tracker->AddRefVar(var_id_))
      return 0;
  } else {
    var_id_ = tracker->AddVar(this);
    if (!var_id_)
      return 0;
  }
  return var_id_;
}

// StringVar -------------------------------------------------------------------

StringVar::StringVar(PP_Module module, const char* str, uint32 len)
    : Var(module),
      value_(str, len) {
}

StringVar::~StringVar() {
}

StringVar* StringVar::AsStringVar() {
  return this;
}

PP_Var StringVar::GetPPVar() {
  int32 id = GetOrCreateVarID();
  if (!id)
    return PP_MakeNull();

  PP_Var result;
  result.type = PP_VARTYPE_STRING;
  result.value.as_id = id;
  return result;
}

// static
PP_Var StringVar::StringToPPVar(PP_Module module, const std::string& var) {
  return StringToPPVar(module, var.c_str(), var.size());
}

// static
PP_Var StringVar::StringToPPVar(PP_Module module,
                                const char* data, uint32 len) {
  scoped_refptr<StringVar> str(new StringVar(module, data, len));
  if (!str || !IsStringUTF8(str->value()))
    return PP_MakeNull();
  return str->GetPPVar();
}

// static
scoped_refptr<StringVar> StringVar::FromPPVar(PP_Var var) {
  if (var.type != PP_VARTYPE_STRING)
    return scoped_refptr<StringVar>();
  scoped_refptr<Var> var_object(
      TrackerBase::Get()->GetVar(static_cast<int32>(var.value.as_id)));
  if (!var_object)
    return scoped_refptr<StringVar>();
  return scoped_refptr<StringVar>(var_object->AsStringVar());
}

}  // namespace ppapi

