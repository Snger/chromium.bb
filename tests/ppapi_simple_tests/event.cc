// Copyright (c) 2011 The Native Client Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <stdint.h>
#include <stdlib.h>

#include <sstream>
#include <string>
#include <queue>

#include "native_client/src/third_party/ppapi/c/pp_input_event.h"
#include "native_client/src/third_party/ppapi/cpp/instance.h"
#include "native_client/src/third_party/ppapi/cpp/module.h"
#include "native_client/src/third_party/ppapi/cpp/var.h"

using std::string;
using std::queue;
using std::ostringstream;

const int kDefaultEventBufferSize = 10;

namespace {

string ModifierToString(uint32_t modifier) {
  string s;
  if (modifier & PP_INPUTEVENT_MODIFIER_SHIFTKEY) {
    s += "shift ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CONTROLKEY) {
    s += "ctrl ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ALTKEY) {
    s += "alt ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_METAKEY) {
    s += "meta ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISKEYPAD) {
    s += "keypad ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISAUTOREPEAT) {
    s += "autorepeat ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_LEFTBUTTONDOWN) {
    s += "left-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_MIDDLEBUTTONDOWN) {
    s += "middle-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_RIGHTBUTTONDOWN) {
    s += "right-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CAPSLOCKKEY) {
    s += "caps-lock ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_NUMLOCKKEY) {
    s += "num-lock ";
  }
  return s;
}

string MouseButtonToString(PP_InputEvent_MouseButton button) {
  switch (button) {
    case PP_INPUTEVENT_MOUSEBUTTON_NONE:
      return "None";
    case PP_INPUTEVENT_MOUSEBUTTON_LEFT:
      return "Left";
    case PP_INPUTEVENT_MOUSEBUTTON_MIDDLE:
      return "Middle";
    case PP_INPUTEVENT_MOUSEBUTTON_RIGHT:
      return "Right";
    default:
      ostringstream stream;
      stream << "Unrecognized (" << static_cast<int32_t>(button) << ")";
      return stream.str();
  }
}

string KeyEvent(PP_InputEvent_Key key,
                PP_TimeTicks time,
                const string& kind) {
  ostringstream stream;
  stream << "Key event:" << kind
         << " modifier:" << ModifierToString(key.modifier)
         << " key_code:" << key.key_code
         << " time:" << time
         << "\n";
  return stream.str();
}

string MouseEvent(PP_InputEvent_Mouse mouse_event,
                PP_TimeTicks time,
                const string& kind) {
  ostringstream stream;
  stream << "Mouse event:" << kind
         << " modifier:" << ModifierToString(mouse_event.modifier)
         << " button:" << MouseButtonToString(mouse_event.button)
         << " x:" << mouse_event.x
         << " y:" << mouse_event.y
         << " click_count:" << mouse_event.click_count
         << " time:" << time
         << "\n";
  return stream.str();
}

string CharEvent(PP_InputEvent_Character char_event, PP_TimeTicks time) {
  ostringstream stream;
  stream << "Character event."
         << " modifier:" << ModifierToString(char_event.modifier)
         << " text:" << char_event.text
         << " time:" << time
         << "\n";
  return stream.str();
}


string WheelEvent(PP_InputEvent_Wheel wheel_event, PP_TimeTicks time) {
  ostringstream stream;
  stream << "Wheel event."
         << " modifier:" << ModifierToString(wheel_event.modifier)
         << " deltax:" << wheel_event.delta_x
         << " deltay:" << wheel_event.delta_y
         << " wheel_ticks_x:" << wheel_event.wheel_ticks_x
         << " wheel_ticks_y:" << wheel_event.wheel_ticks_y
         << " scroll_by_page:"
         << (wheel_event.scroll_by_page ? "true" : "false")
         << "\n";
  return stream.str();
}

string EventToString(const PP_InputEvent& event) {
  ostringstream stream;
  switch (event.type) {
    default:
    case PP_INPUTEVENT_TYPE_UNDEFINED:
      stream << "Unrecognized Event (" << static_cast<int32_t>(event.type)
             << ")";
      return stream.str();

    case PP_INPUTEVENT_TYPE_MOUSEDOWN:
      return MouseEvent(event.u.mouse, event.time_stamp, "Down");
    case PP_INPUTEVENT_TYPE_MOUSEUP:
      return MouseEvent(event.u.mouse, event.time_stamp, "Up");
    case PP_INPUTEVENT_TYPE_MOUSEMOVE:
      return MouseEvent(event.u.mouse, event.time_stamp, "Move");
    case PP_INPUTEVENT_TYPE_MOUSEENTER:
      return MouseEvent(event.u.mouse, event.time_stamp, "Enter");
    case PP_INPUTEVENT_TYPE_MOUSELEAVE:
      return MouseEvent(event.u.mouse, event.time_stamp, "Leave");
    case PP_INPUTEVENT_TYPE_MOUSEWHEEL:
      return WheelEvent(event.u.wheel, event.time_stamp);
    case PP_INPUTEVENT_TYPE_RAWKEYDOWN:
      return KeyEvent(event.u.key, event.time_stamp, "RawKeyDown");
    case PP_INPUTEVENT_TYPE_KEYDOWN:
      return KeyEvent(event.u.key, event.time_stamp, "Down");
    case PP_INPUTEVENT_TYPE_KEYUP:
      return KeyEvent(event.u.key, event.time_stamp, "Up");
    case PP_INPUTEVENT_TYPE_CHAR:
      return CharEvent(event.u.character, event.time_stamp);
  }
}

void StringReplace(string* input,
                   const string& find,
                   const string& replace) {
  if (find.length() == 0 || input->length() == 0) {
    return;
  }

  size_t start_pos = 0;
  while (1) {
    start_pos = input->find(find, start_pos);
    if (start_pos == string::npos) {
      break;
    }
    input->replace(start_pos, find.length(), replace);
    start_pos += replace.length();
  }
}

}  // namespace


class MyInstance : public pp::Instance {
 private:
  size_t max_buffer_size_;
  queue<PP_InputEvent> event_buffer_;

  void ParseArgs(uint32_t argc, const char* argn[], const char* argv[]) {
     for (uint32_t i = 0; i < argc; ++i) {
       const std::string tag = argn[i];
       if (tag == "buffer_size") max_buffer_size_ = strtol(argv[i], 0, 0);
       // ignore other tags
     }
  }

  // Dump all the event via PostMessage for testing
  void FlushEventBuffer() {
      while (event_buffer_.size() > 0) {
        ostringstream stream;
        stream << pp_instance() << ": " << EventToString(event_buffer_.front());
        event_buffer_.pop();
        // Replace space with underscore to simplify testing
        string s = stream.str();
        StringReplace(&s, " ", "_");
        pp::Var message(s);
        PostMessage(message);
      }
  }

 public:
  explicit MyInstance(PP_Instance instance)
    : pp::Instance(instance), max_buffer_size_(kDefaultEventBufferSize) {}

  virtual ~MyInstance() {}

  virtual bool Init(uint32_t argc, const char* argn[], const char* argv[]) {
    ParseArgs(argc, argn, argv);
    return true;
  }

  virtual bool HandleInputEvent(const PP_InputEvent& event) {
    event_buffer_.push(event);
    if (event_buffer_.size() >= max_buffer_size_) {
      FlushEventBuffer();
    }
    return true;
  }

  virtual void DidChangeFocus(bool has_focus) {
    FlushEventBuffer();
  }
};

// standard boilerplate code below
class MyModule : public pp::Module {
 public:
  virtual pp::Instance* CreateInstance(PP_Instance instance) {
    return new MyInstance(instance);
  }
};

namespace pp {
  Module* CreateModule() {
    return new MyModule();
  }
}
