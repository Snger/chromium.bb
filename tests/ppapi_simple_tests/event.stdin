# define a bunch of variables for readability

set_variable MODULE i(4444)
set_variable INSTANCE i(5555)

# From ppapi/c/pp_input_event.h
#
# struct PP_InputEvent {
#   PP_InputEvent_Type type;
#   int32_t padding;
#   PP_TimeTicks time_stamp;  // double
#   union {
#     struct PP_InputEvent_Key key;
#     struct PP_InputEvent_Character character;
#     struct PP_InputEvent_Mouse mouse;
#     struct PP_InputEvent_Wheel wheel;
#     char padding[64];
#   } u;
# };
#
# struct PP_InputEvent_Mouse {
#   uint32_t modifier;
#   PP_InputEvent_MouseButton button;
#   float x;
#   float y;
#   int32_t click_count;
#  }; // 20 bytes -> 44 bytes paddding
#
#
# struct PP_InputEvent_Key {
#   uint32_t modifier;
#   uint32_t key_code;
# };  // 8 bytes -> 56 bytes padding

set_variable ZERO_44 8:0,8:0,8:0,8:0,8:0,4:0
set_variable ZERO_56 8:0,8:0,8:0,8:0,8:0,8:0,8:0

set_variable TIME_1 8:0x3ff0000000000000
set_variable TIME_2 8:0x3ff1000000000000
set_variable TIME_3 8:0x3ff2000000000000
set_variable TIME_4 8:0x3ff3000000000000
set_variable TIME_5 8:0x3ff4000000000000
set_variable TIME_6 8:0x3ff5000000000000
set_variable TIME_7 8:0x3ff6000000000000

set_variable COORDS 4:0x43840000,4:0x42c80000

set_variable EV_MOUSE_DOWN  R(80,4:0,4:0,${TIME_1},8:64,${COORDS},4:1,${ZERO_44})
set_variable EV_MOUSE_UP    R(80,4:1,4:0,${TIME_2},8:64,${COORDS},4:0,${ZERO_44})

set_variable EV_MOUSE_MOVE  R(80,4:2,4:0,${TIME_3},8:0,${COORDS},4:0,${ZERO_44})

set_variable EV_MOUSE_ENTER R(80,4:3,4:0,${TIME_4},8:0,${COORDS},4:0,${ZERO_44})
set_variable EV_MOUSE_LEAVE R(80,4:4,4:0,${TIME_5},8:0,${COORDS},4:0,${ZERO_44})

set_variable EV_KEY_DOWN    R(80,4:7,4:0,${TIME_6},4:0,4:0x54,${ZERO_56})
set_variable EV_KEY_UP      R(80,4:8,4:0,${TIME_7},4:0,4:0x54,${ZERO_56})

# TODO(robertm) figure out what this represents, maybe PP_Var
set_variable PROPERTY C(16,\x05\x00\x00\x00\x06\x00\x00\x00events\x00\x00)

# TODO(robertm) figure out what this represents, maybe PP_Var
set_variable EXCEPTION R(8,8:0)


echo
echo "*** INIT MODULE"
replay 1 PPB_GetInterface:s:i  s("PPB_Core;0.5") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_Instance;1.0") * i(1)
# the results contain a process id which is non-determinisitic
rpc PPP_InitializeModule hide-results i(0) ${MODULE} h(pepper_desc) s("${service_string}") * i(0) i(0)

echo
echo "*** INIT INSTANCE"
set_variable TAGS C(12,buffer_size\000)
set_variable VALUES C(4,100\000)
rpc PPP_Instance_DidCreate ${INSTANCE} i(1) ${TAGS} ${VALUES} * i(0)

echo
echo "*** SENT EVENTS EV_MOUSE_DOWN"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_DOWN} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_UP"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_UP} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_MOVE"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_MOVE} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_ENTER"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_ENTER} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_LEAVE"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_MOUSE_LEAVE} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_UP"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_KEY_UP} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_DOWN"
rpc PPP_Instance_HandleInputEvent ${INSTANCE} ${EV_KEY_DOWN} * i(0)


echo
echo "*** TRIGGER EVENT FLUSH"
replay 1 PPB_GetInterface:s:i  s("PPB_Var;1.0") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_Messaging;1.0") * i(1)
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(104,\x05\x00\x00\x00_\x00\x00\x005555:_Mouse_event:Down_modifier:left-button-down__button:Left_x:264_y:100_click_count:1_time:1\n\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(112,\x05\x00\x00\x00b\x00\x00\x005555:_Mouse_event:Up_modifier:left-button-down__button:Left_x:264_y:100_click_count:0_time:1.0625\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Move_modifier:_button:Left_x:264_y:100_click_count:0_time:1.125\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(96,\x05\x00\x00\x00T\x00\x00\x005555:_Mouse_event:Enter_modifier:_button:Left_x:264_y:100_click_count:0_time:1.1875\n\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Leave_modifier:_button:Left_x:264_y:100_click_count:0_time:1.25\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(64,\x05\x00\x00\x004\x00\x00\x005555:_Key_event:Up_modifier:_key_code:84_time:1.375\n\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555)  C(64,\x05\x00\x00\x007\x00\x00\x005555:_Key_event:Down_modifier:_key_code:84_time:1.3125\n\x00) *
rpc PPP_Instance_DidChangeFocus ${INSTANCE} b(0) *


rpc PPP_ShutdownModule *

echo "DONE"
