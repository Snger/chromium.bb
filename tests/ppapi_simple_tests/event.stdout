
"*** REPLAY PROLOG"

"*** REGISTER REPLAY METHODS"

"*** CREATE UPCALL SERVICE"

"*** INIT MODULE"
rpc call intiated PPP_InitializeModule:iihs:ii
input 0:  i(0)
input 1:  i(4444)
input 2:  h(pepper_desc)
input 3:  @LENGTHY_SERVICE_STRING
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Core;0.5")
output 0:  i(1)
rpc call complete PPP_InitializeModule:iihs:ii


"*** INIT INSTANCE"
rpc call intiated PPP_Instance_DidCreate:iiCC:i
input 0:  i(5555)
input 1:  i(1)
input 2:  C(12,buffer_size\x00)
input 3:  C(4,100\x00)
rpc call complete PPP_Instance_DidCreate:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_DOWN"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_UP"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf1?@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_MOVE"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf2?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_ENTER"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf3?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_LEAVE"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf4?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x84C\x00\x00\xc8B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_UP"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf6?\x00\x00\x00\x00T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_DOWN"
rpc call intiated PPP_Instance_HandleInputEvent:iC:i
input 0:  i(5555)
input 1:  C(80,\x07\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf5?\x00\x00\x00\x00T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
rpc call complete PPP_Instance_HandleInputEvent:iC:i
output 0:  i(1)


"*** TRIGGER EVENT FLUSH"
rpc call intiated PPP_Instance_DidChangeFocus:ib:
input 0:  i(5555)
input 1:  b(0)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Var;0.5")
output 0:  i(1)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Messaging;0.1")
output 0:  i(1)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Var(Deprecated\x29;0.3")
output 0:  i(1)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(104,\x05\x00\x00\x00_\x00\x00\x005555:_Mouse_event:Down_modifier:left-button-down__button:Left_x:264_y:100_click_count:1_time:1\n\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(112,\x05\x00\x00\x00b\x00\x00\x005555:_Mouse_event:Up_modifier:left-button-down__button:Left_x:264_y:100_click_count:0_time:1.0625\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Move_modifier:_button:Left_x:264_y:100_click_count:0_time:1.125\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00T\x00\x00\x005555:_Mouse_event:Enter_modifier:_button:Left_x:264_y:100_click_count:0_time:1.1875\n\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Leave_modifier:_button:Left_x:264_y:100_click_count:0_time:1.25\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(64,\x05\x00\x00\x004\x00\x00\x005555:_Key_event:Up_modifier:_key_code:84_time:1.375\n\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(64,\x05\x00\x00\x007\x00\x00\x005555:_Key_event:Down_modifier:_key_code:84_time:1.3125\n\x00)
rpc call complete PPP_Instance_DidChangeFocus:ib:

rpc call intiated PPP_ShutdownModule::
rpc call complete PPP_ShutdownModule::

"DONE"
